File: reverseOnlyLetters.kt
class Solution {
    fun reverseOnlyLetters(s: String): String {
        val n = s.length
        val arr = s.toCharArray()
        var left = 0
        var right = n - 1
        while (true) {
            while (left < right && !Character.isLetter(s[left])) { // 判断左边是否扫描到字母
                left++
            }
            while (right > left && !Character.isLetter(s[right])) { // 判断右边是否扫描到字母
                right--
            }
            if (left >= right) {
                break
            }
            swap(arr, left, right)
            left++
            right--
        }
        return String(arr)
    }
    fun swap(arr: CharArray, left: Int, right: Int) {
         arr[left] = arr[right].also {
            arr[right] = arr[left]
        }
        /*
         val temp = arr[left]
        arr[left] =  arr[right]
        arr[right] = temp
        */
    }
}

File: reverseOnlyLetters.c
static inline void swap(char *l, char *r) {
    *l^= *r;
    *r^= *l;
    *l^= *r;
}
char *reverseOnlyLetters(char *s){
    int n = strlen(s);
    int left = 0, right = n - 1;
    while (true) {
        while (left < right && !isalpha(s[left])) { // 判断左边是否扫描到字母
            left++;
        }
        while (right > left && !isalpha(s[right])) { // 判断右边是否扫描到字母
            right--;
        }
        if (left >= right) {
            break;
        }
        swap(s + left, s + right);
        left++;
        right--;
    }
    return s;
}


